New page to draw random cards

ChristophNiehoff opened the pull request.

ghost91- commented on apps/client/src/components/randomcarddisplay/randomCardDisplay.tsx:
```diff
@@ -0,0 +1,87 @@
+          />
+        </div>
+      )}
+      <Button block size="lg" color="primary" onClick={() => selectCard()}>
```
```suggestion
      <Button block size="lg" color="primary" onClick={selectCard}>
```

ChristophNiehoff replied:
Done

ghost91- commented on apps/client/src/components/randomcarddisplay/randomCardDisplay.tsx:
```diff
@@ -0,0 +1,87 @@
+
+const suites: Suit[] = ['A', 'B', 'C', 'D', 'E', 'T'];
+
+const RandomCardDisplay: FC = () => {
+  const allDecks = Object.keys(CARD_DECKS) as GameMode[];
+  const [selectedDecks, setSelectedDecks] = useState<GameMode[]>(allDecks);
+  const [selectedCard, setSelectedCard] = useState<SelectableCard | undefined>(
+    undefined,
+  );
+  const selectableCards: SelectableCard[] = [];
+
+  selectedDecks.forEach((deck) =>
+    suites.forEach((suit) => {
+      selectableCards.push(
+        ...CARD_DECKS[deck][suit].cards.map((card) => ({
+          card,
+          gameMode: deck,
+        })),
+      );
+    }),
+  );
+
+  const selectCard = () =>
+    setSelectedCard(
+      selectableCards[Math.floor(Math.random() * selectableCards.length)],
+    );
+
+  useEffect(() => {
+    selectCard();
+  }, []);
+
+  const toggleCardDeck = (deck: GameMode) => {
+    if (selectedDecks.includes(deck)) {
+      setSelectedDecks(
+        selectedDecks.filter((selectedDeck) => selectedDeck !== deck),
+      );
+    } else {
+      setSelectedDecks(selectedDecks.concat(deck));
+    }
+  };
```
```suggestion
const suites: Suit[] = ['A', 'B', 'C', 'D', 'E', 'T'];
const allDecks = Object.keys(CARD_DECKS) as GameMode[];

const RandomCardDisplay: FC = () => {
  const [selectedDecks, setSelectedDecks] = useState<GameMode[]>(allDecks);

  const selectableCards = selectedDecks.flatMap((deck) =>
    suites.flatMap((suit) =>
      CARD_DECKS[deck][suit].cards.map((card) => ({
        card,
        gameMode: deck,
      })),
    ),
  );

  const getRandomSelectableCard = () =>
    selectableCards[Math.floor(Math.random() * selectableCards.length)];

  const [selectedCard, setSelectedCard] = useState(getRandomSelectableCard());

  const selectCard = () => setSelectedCard(getRandomSelectableCard());

  const toggleCardDeck = (deck: GameMode) =>
    setSelectedDecks((selectedDecks) =>
      selectedDecks.includes(deck)
        ? selectedDecks.filter((selectedDeck) => selectedDeck !== deck)
        : [...selectedDecks, deck],
    );
```

ghost91- commented on apps/client/src/components/randomcarddisplay/randomCardDisplay.tsx:
```diff
@@ -0,0 +1,87 @@
+import { Card, CARD_DECKS, GameMode, Suit } from '@eop/shared';
+import { FC, useEffect, useState } from 'react';
+import DealtCard from '../dealtcard/dealtcard';
+import './randomCardDisplay.css';
```
I know, we haven't been doing it so far, but I would suggest to start using CSS modules for new styles, instead of global CSS. In order to do so here, the following changes are needed:
* Rename `randomCardDisplay.css` to `randomCardDisplay.module.css`
* Change this import to `import classes from './randomCardDisplay.module.css';`
* Reference the classes, e.g., instead of `className="card-container"`, we would use `className={classes['card-container']}`

ChristophNiehoff replied:
done.

ghost91- commented on apps/client/src/pages/random-card.tsx:
I would suggest to include some way to navigate back to the create page

ChristophNiehoff replied:
This happens when clicking the EoP logo. For me, that is enough.

ghost91- commented on apps/client/src/pages/create.tsx:
```diff
@@ -271,19 +270,20 @@ class Create extends React.Component<CreateProps, CreateState> {
-        <small className="text-secondary">
-          To start playing, select the number of players and enter their names.
-        </small>
+        <Button
+          block
+          size="lg"
+          color="primary"
+          onClick={() => (window.location.href = `/random-card`)}
+        >
+          Draw a random card
+        </Button>
         <hr />
+        <p className="space-top">
+          <center>
+            <strong>Or create a new game:</strong>
+          </center>
+        </p>
```
To be honest, I don't quite like how this looks. In my opinion, this makes the "Draw a random card" button much too prominent. Playing the games is the primary function of this application, so doing the setup for that should be the primary action, here, not going to a separate page where you can look at the cards. What is the exact intention of that functionality?

ChristophNiehoff replied:
Done.

ghost91- commented on apps/client/src/pages/create.tsx:
```diff
@@ -271,19 +270,20 @@ class Create extends React.Component<CreateProps, CreateState> {
+        </Button>
         <hr />
+        <p className="space-top">
+          <center>
+            <strong>Or create a new game:</strong>
+          </center>
```
This should not be done with HTML elements. It's just styling, so it should just be CSS.

ChristophNiehoff replied:
Becomes obsolete with above change

ghost91- commented on apps/client/src/pages/random-card.tsx:
```diff
@@ -0,0 +1,41 @@
+    <div>
+      <Banner />
+      {/* @ts-expect-error This seems to be incorrectly typed in helmet*/}
+      <Helmet bodyAttributes={{ style: 'background-color : #000' }} />
```
See https://github.com/TNG/elevation-of-privilege/pull/137

ChristophNiehoff replied:
done

ChristophNiehoff commented on apps/client/src/components/randomcarddisplay/randomCardDisplay.tsx:
```diff
@@ -0,0 +1,87 @@
+      </FormGroup>
+      {selectedCard && (
+        <div className="card-container">
+          <DealtCard
```
Fixed this issue within CSS, rather than using an awkward shifting of the of EoMLSec cards in `DealtCard`.

ChristophNiehoff commented:
@ghost91- I addressed your comments.
However, #137 introduced a "bug": Via the global CSS'ing now every background was black (also when playing the game).
Instead of fixing all the global CSS, I took this as an opportunity, to make the background white everywhere. What do you think?

(current version is deployed to staging env.)

ghost91- commented on apps/client/src/styles/about.css:
```diff
@@ -1,3 +1,3 @@
 html body {
-  background-color: #000;
+  background-color: #fff;
```
if we want the background to be white on _all_ pages, we should probably put it into an `index.css` file that we import from index.tsx.

ghost91- reviewed (changes requested):
Just one more tiny comment, otherwise this looks fine. Let me know when the changes are done (or if I should do them myself), then we can merge this ASAP.
